src/trust.rs

```diff
@@ -118,6 +118,9 @@ pub fn create_device_certificate(
     let device_csr = generate_certificate(CertificateType::device, &device_id, &app_id, days)?;
     let ca_cert_fin = Certificate::from_params(ca_certificate)?;
 
+    // Checking equality of public keys of Cert from application object and supplied CA key
+    verify_public_key(ca_cert_pem, &ca_cert_fin.serialize_der()?)?;
```

Comment:
So `ca_cert_perm` comes from the `app` object (pulled from drogue-cloud API) and you compare it to `ca_cert_fin` which was reconstructed from the `ca_cert_pem` + the private key the user passed as input. 
If the key was not matching the cert, why `CertificateParams::from_ca_cert_pem(&ca_cert_pem, ca_key_content)` would not fail ? 
Thanks in advance for the explanation :)

Reply:
> If the key was not matching the cert, why CertificateParams::from_ca_cert_pem(&ca_cert_pem, ca_key_content) would not fail ?

I agree with you, but there is no such validation in the `rcgen` crate. In the `rcgen` crate, it uses the certificate to get subj and issuer information. No public key validation is done.
https://github.com/est31/rcgen/blob/master/src/lib.rs#L632

In community messages, rcgen guys want us to first use [verify_signature](https://docs.rs/x509-parser/0.9.2/x509_parser/certificate/struct.X509Certificate.html#method.verify_signature) before using `from_ca_cert_pem`

But in your case, we have to do many conversions just to create the right parameters to pass to `verify_signature` so it's okay to just validate the public key.

Reply:
Ok thanks